rfd/0002-custom-terraform-registry.md

```diff
@@ -107,7 +161,149 @@ We _could_ modify Houston to serve the required files, but as the service is
+
+For example, this discovery protocol file:
+
+```json
+{
+    "providers.v1": "/registry/"
+}
+```
+
+_**NOTE:**_ The address in the discovery protocol JSON need not be a relative
+path - it could _in theory_ point to a totally different host. It _should_ be
+possible to have the discovery protocol served by `goteleport.com` that would
+direct terraform to `terraform.releases.teleport.dev` for the actual registry,
+if that's the preferred branding.
+
+Note that I have not _personally_ tested this behaviour.
+
+The practical change this would make would be that the provider source would
+change from (in the example above):
+
+```terraform
+source = "terraform.releases.teleport.dev/gravitational/teleport"
+```
+
+to
+
+```terraform
+source = "goteleport.com/gravitational/teleport"
+```
+
+##### 2. The registry (key prefix `/registry`)
+
+This is where the main registry metadata lives. The basic structure is described
+in the [Provider Registry Specification](https://www.terraform.io/internals/provider-registry-protocol),
+but in short there are two parts of the index for each `namespace/provider`
+pair: the `versions` index, and the `download` metadata.
+
+For our purposes, with a namespace of `gravitational` and a provider name of
+`teleport`, the `versions` index file will be located at `/registry/gravitational/teleport/versions`
+
+This is the only object in the repository that need modification as part of
+normal use.
+
+The `download` metadata for each version/OS/architecture triple is stored
+according to the rule
+
+```shell
+/registry/gravitational/teleport/:version/:os/:architecture
+```
+
+Once created, each `download` metadata does not need modification, unless
+a provider object is re-published.
+
+##### 3. The object store (key prefix `/store`)
+
+The object store is where all the release package files are stored. Each
+provider artefact, together with its associated hash & signatures files.
+
+This section of the bucket should be write-only. No objects in this section
+should be modifiable or delete-able.
+
+#### Build trigger and inputs
+
+By analogy with the main Teleport release system, the Terraform registry
+packaging and publishing will be happen during release _promotion_ task on
+Drone.
+
+##### Release Tarball
+
+The Teleport provider release process currently produces a compressed tarball
+by default, and this should still be considered the primary artefact product
+of the _release_ build, at least as long as Houston is expected to serve
+downloads of the Terraform provider.
+
+The release tarball(s) will be injected into the build by being downloaded
+from the staging bucket, based on the version tag supplied during the
+promotion process.
+
+##### Terraform Plugin API Version
+
+In order to correctly index the providers, the registry need to know what
+version of the Terraform Plugin API a given provider supports. This is not
+something we can deduce from the provider itself, and must be "just known"
+in advance.
```

Comment:
What do these versions look like and where do they come from?  Is it possible that info is encoded in the go.mod or makefile somehow, and our release automation could derive it from there, such that it is always correct with no update required (outside of bumping the go mod or whatever).

Reply:
Nope. You apparantley just gotta know it.

Reply:
Longer version: Ultimately the API version numbers come from some constants buried deep inside the the transitive dependencies of the Terraform plugin framework. Terraform uses GRPC as the basis of it's plugin API, so they are  probably originally specified in a protobuf file somewhere.

All of the the types and code that Terraform uses to load and validate plugins are inside `internal` packages used by the Terraform command-line app, and so are not easily available for us to just pick up and use. There _may_ be a way to do it by subverting the testing hashicorp-supplied Provider testing framework, but that's getting really dodgy. 

At the moment no automated solution is looking less bad than manually making sure the dronefile injects the right versions into the promotion code.

Reply:
I'd add that those API version upgrades are really rare and always backwards-compatible, as far as I understood. We even might want to stick with an older version for a while when this upgrade happens, to make things work with older Terraform versions.